docs/rfcs/025-generation-numbers.md

```diff
@@ -0,0 +1,468 @@
+- 16 bit node ID (this doesn't have to accommodate all pageservers ever, just all the pageservers
+  that act together in a logical cluster)
+- 32 bit node generation (as with attachment generation, this can handle even very frequent restarts)
+
+So an example suffix for generation 1 of node 0 might look like:
+
+```
+   00000001-0000-00000001
+```
+
+This suffix provides a lot of flexibility in how we may handle migration/restarts/failover
+at runtime:
+
+- We can replace a pageserver with another physical node having the same pageserver ID, and
+  the pageserver ID in the name will protect from conflicting writes.
+- Two pageservers may have the same tenant attached without risking stepping on each others
+  writes.
+- Ultimately, pageservers might self-coordinate such that a single "attachment" refers to
+  multiple pageservers, and they may internally do HA without troubling the control
+  plane to do new attachments, as the node generation makes this safe.
+
+### Deletion Changes
+
+While writes are de-conflicted by writers always using their own generation number in the key,
+deletions are slightly more challenging: if a pageserver A is isolated, and the true active node is
+pageserver B, then it is dangerous for A to do any object deletions, even of objects that it wrote
+itself, because pageserver's B metadata might reference those objects.
+
+To solve this without doing excessive synchronization, deletions are accumulated in deletion lists
+(also known as deadlists, or garbage lists), to be executed later. The actual execution of one of
+these lists has two preconditions:
+
+- That the executing node has written out the metadata index file since it decided to do
+  the deletion (i.e. no to-be-deleted files are referenced by the index)
+- That after writing that index file, the deleting node has confirmed that it is still the latest
+  generation (i.e. none of its attached tenants have been re-attached to another node).
+
+The order of operations for correctness is:
+
+1. Decide to delete one or more objects
+2. Write updated metadata (this means that if we are still the rightful leader, any subsequent leader will
+   read metadata that does not reference the about-to-be-deleted object).
+3. Confirm we hold a non-stale generation (this validates the assumption in step 2, and taken
+   together, these provide a guarantee that no future leader will attempt to read the objects
+   we are about to delete).
+4. Actually delete the objects
+
+Because step 3 puts load on the control plane, deletion lists should be executed very lazily: write
+them out to S3 and then execute them as a low priority batch operation in the background.
+
+Deletion lists have an a couple of additional benefits:
+
+- efficiency: accumulate enough objects
+  to populate plural DeleteObjects requests, rather than issuing many smaller `DeleteObject` requests.
+- read replicas: if some other node is reading data in S3 via some older metadata, the longer we delay
```

Comment:
I'm thinking about the general case, including compaction, where a read replica may be reading from some totally valid LSN, but the underlying S3 object it tries to read might have been deleted by compaction on the master.

This is a safe scenario, because a read replica can take that 404 as a hint to go re-read the latest metadata and retry the read, but delaying the object deletions can make those kinds of stalls on readers vanishingly unlikely, which is nice for performance.